ElectronClient/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearch.ts

```diff
@@ -0,0 +1,151 @@
+import { useCallback, useEffect, useState } from 'react';
+// Helper functions that use the cursor
+export default function useCursorUtils(CodeMirror: any) {
+
+	const [markers, setMarkers] = useState([]);
+	const [overlay, setOverlay] = useState(null);
+	const [scrollbarMarks, setScrollbarMarks] = useState(null);
+	const [overlayTimeout, setOverlayTimeout] = useState(null);
+	const [previousKeywordValue, setPreviousKeywordValue] = useState(null);
+
+	const clearMarkers = useCallback(() => {
+		for (let i = 0; i < markers.length; i++) {
+			markers[i].clear();
+		}
+
+		setMarkers([]);
+	}, [markers]);
```

Comment:
useCallback is not needed here because `defineExtension()` will capture in the closure the first version of clearMarkers it sees, so even if it's updated when `markers` changes, it won't update it in the closure.

In this particular case, I think you can have a regular function and clear the markers like so:

```js
function clearMarkers() {
	setMarkers(function(previous) {
		for (let i = 0; i < previous.length; i++) {
			previous[i].clear();
		}
		return [];
	});
}
```



The nice thing with [functional updates](https://reactjs.org/docs/hooks-reference.html#functional-updates) is that you're guaranteed that you won't use a stale version of the state. The state setter will always pass the latest version of the state value to the function.

Reply:
Thanks for this, and the tip about using a reducer. I learned a lot about hooks and with all this struggle i think I'm a lot more comfortable with them. 
In the end I didn't use functional updates or a reducer pattern, I just reverted this code back to what it was previously (I didn't have an issues with stale states before so I never should have changed this). I think the reason that the states never go stale is because changing states triggers a "re-render" which redefines all the functions including the extension, so the next time this gets called, all the states are up-to-date as far as the extension is concerned. The only stale issue was in `useEffect` because it can be called outside of the extension code (on unmounts) so it needs the extra guard.